Enable generation of llvm.lifetime.start/.end intrinsics

rkarrenberg opened the pull request:
Enable generation of llvm.lifetime.start/.end intrinsics.

- Remove HLSL change from CGDecl.cpp::EmitLifetimeStart() that disabled generation of lifetime markers in the front end.
- Enable generation of lifetime intrinsics when inlining functions (PassManagerBuilder.cpp).
- Both of these cover a different set of situations that can lead to inefficient code without lifetime intrinsics (see examples below):
  - Assume a struct is created inside a loop but some or all of its fields are only initialized conditionally before the struct is being
    used. If the alloca of that struct does not receive lifetime intrinsics before being lowered to SSA its definition will effectively be
    hoisted out of the loop, which changes the original semantics: Since the initialization is conditional, the correct SSA form for this
    code requires a phi node in the loop header that persists the value of the struct field throughout different iterations because
    the compiler doesn't know anymore that the field can't be initialized in a different iteration than when it is used.
  - If the lifetime of an alloca in a function is the entire function it doesn't need lifetime intrinsics. However, when inlining that
    function, the alloca's lifetime will then suddenly span the entire caller, causing similar issues as described above.
- For backwards compatibility, replace lifetime.start/.end intrinsics with a store of undef in DxilPreparePasses.cpp, or, for validator
  version < 1.6, with a store of 0 (undef store is disallowed). This is slightly inconvenient but achieves the same goal as the lifetime
  intrinsics. The zero initialization is actually the current manual workaround for developers that hit one of the above issues.
- Allow lifetime intrinsics to pass DXIL validation.
- Allow undef stores to pass DXIL validation.
- Allow bitcast to i8* to pass DXIL validation.
- Make various places in the code aware of lifetime intrinsics and their related bitcasts to i8*.
- Adjust ScalarReplAggregatesHLSL so it generates new intrinsics for each element once a structure is broken up. Also make sure
  that lifetime intrinsics are removed when replacing one pointer by another upon seeing a memcpy. This is required to prevent a
  pointer accidentally "inheriting" wrong lifetimes.
- Adjust PromoteMemoryToRegister to treat an existing lifetime.start intrinsic as a definition.
- Since lifetime intrinsics require a cleanup, the logic in CGStmt.cpp:EmitBreakStmt() had to be changed:
  EmitHLSLCondBreak() now returns the generated BranchInst. That branch is then passed into EmitBranchThroughCleanup(),
  which uses it instead of creating a new one. This way, the cleanup is generated correctly and the wave handling also still works
  as intended.
- Adjust a number of tests that now behave slightly differently. memcpy_preuser.hlsl was actually exhibiting exactly the situation
  explained above and relied on the struct definition of "oStruct" to be hoisted out to produce the desired IR. And entry_memcpy()
  in cbuf_memcpy_replace.hlsl required an explicit initialization: With lifetime intrinsics, the original code correctly collapsed to
  returning undef. Without lifetime intrinsics, the compiler could not prove this. With proper initialization, the test now has the
  intended effect, even though the collapsing to undef could be a desireable test for lifetime intrinsics.

Example 1:

Original code:
```
for( ;; ) {
  func();
  MyStruct s;
  if( c ) {
    s.x = ...;
    ... = s.x;
  }
  ... = s.x;
}
```
Without lifetime intrinsics, this is equivalent to:
```
MyStruct s;
for( ;; ) {
  func();
  if( c ) {
    s.x = ...;
    ... = s.x;
  }
  ... = s.x;
}
```
After SROA, we now have a value live across the function call, which will cause a spill:
```
for( ;; ) {
  x_p = phi( undef, x_p2 );
  func();
  if( c ) {
    x1 = ...;
    ... = x1;
  }
  x_p2 = phi( x_p, x1 );
  ... = x_p2;
}
```

Example 2:
```
void consume(in Data data);
void expensiveComputation();

bool produce(out Data data) {
    if (condition) {
        data = ...; // <-- conditional assignment of out-qualified parameter
        return true;
    }
    return false; // <-- out-qualified parameter left uninitialized
}
void foo(int N) {
    for (int i=0; i<N; ++i) {
        Data data;
        bool valid = produce(data); // <-- generates a phi to prior iteration's value when inlined. There should be none
        if (valid)
            consume(data);
        expensiveComputation(); // <-- said phi is alive here, inflating register pressure
    }
}
```

pow2clk commented on lib/HLSL/HLMatrixLowerPass.cpp:
```diff
@@ -537,6 +545,24 @@ void HLMatrixLowerPass::replaceAllVariableUses(
+        DXASSERT(BCI->getNumUses() == 1, "expected only a single use");
+        llvm::Use &UseUse = *BCI->use_begin();
+        IntrinsicInst *Intrin = dyn_cast<IntrinsicInst>(UseUse.getUser());
+        DXASSERT(Intrin && (Intrin->getIntrinsicID() == Intrinsic::lifetime_start || Intrin->getIntrinsicID() == Intrinsic::lifetime_end), "use must be lifetime.start/end");
```
The simplest way to fix the Clang builds would be to use `DXASSERT_LOCALVAR` here and provide Intrin as the first variable. You could also move the `dyn_cast<IntrinsicInst>(UseUse.getUser())` into the DXASSERT.

rkarrenberg replied:
Fixed.
I couldn't repro the failure, though, is there anything special I have to do? building the ALL_BUILD target in VS in Release mode did not trigger it, at least not in an incremental build.

pow2clk replied:
The failures were in Linux and Apple builds using the clang compiler, which has warnings as errors that don't perfectly overlap with those produced by MSVC, but can help reveal programming mistakes that apply to all platforms.

rkarrenberg replied:
I see, thanks!

tex3d commented on tools/clang/lib/CodeGen/CGStmt.cpp:
```diff
@@ -1202,11 +1202,19 @@ void CodeGenFunction::EmitBreakStmt(const BreakStmt &S) {
+    // We execute this if
+    // - we are in an unnested loop, or
+    // - we are in a nested loop but the continue block of the enclosing loop is different from the current continue block.
+    // TODO: Is the second conditional even possible in a language without goto or "continue N"?
```
I suppose I'd like @pow2clk to comment here.

pow2clk replied:
Correct the second case to be "we are in a nested control *control construct* but the continue block of the enclosing loop is different from the current continue block"

In order for the TODO to make sense to me, I think it's asking how that it is possible that a nested loop have the same continue as the outer loop. It seems trivially easy for a nested loop to have a different continue as the outer loop.

The answer to the quandry I altered so that I have an interesting answer to is: a switch statement (control construct) inside a loop. In this case, since the switch needs a break location, but doesn't have a continue location that is unique from the enclosing loop, clang copies the continue from the containing loop since any continue found within the switch should point to the continue for that outer loop. Previously this confused the implementation and added dx.breaks for switch statements, which wasn't necessary.

I admit I failed to document this properly. There is mention of it in the description of this commit: https://github.com/microsoft/DirectXShaderCompiler/pull/2795/commits/6ea8dc00ec0d662db23a10cfc43d6abf28885eca which is the only way I was able to remember what was going on. I should have added a comment to the code. I was young and foolish when I wrote this.

rkarrenberg replied:
Fixed the comment.
Thanks for the explanation @pow2clk! I missed the fact that this doesn't have to be nested loops but other constructs can have break or continue locations, too.
Removed the TODO, this was purely for my own understanding :). Do you want me to remove the rest of the comment, too, or replace it with more explanation while we're at it?

tex3d commented on lib/HLSL/HLMatrixLowerPass.cpp:
```diff
@@ -537,6 +545,24 @@ void HLMatrixLowerPass::replaceAllVariableUses(
+      // Replace bitcasts to i8* for lifetime intrinsics.
+      if (BCI->getType()->isPointerTy() && BCI->getType()->getPointerElementType()->isIntegerTy(8))
+      {
+        DXASSERT(BCI->getNumUses() == 1, "expected only a single use");
```
Couldn't one bitcast be used for both lifetime.start and lifetime.end in a single block?  Why can it only have one use?

rkarrenberg replied:
Absolutely, though currently there's always one bitcast per intrinsic. Makes sense to be more flexible here. Fixed.

rkarrenberg commented on lib/HLSL/DxilPreparePasses.cpp:
```diff
@@ -376,6 +376,80 @@ class DxilFinalizeModule : public ModulePass {
+      Value *ptr = nullptr;
+      if (BitCastInst *BC = dyn_cast<BitCastInst>(voidPtr)) {
+        ptr = BC->getOperand(0);
+      } else if (ConstantExpr *CE = dyn_cast<ConstantExpr>(voidPtr)) {
```
It does. This can happen if a local variable/array is promoted to a constant global. I think lifetime intrinsics are perfectly fine for constant globals even though they may not be of much value. An example where this happens is the float3 arrays DirLightDirections and DirLightColors in HLSLFileCheck/samples/d3d11/SubD11_SmoothPS.hlsl or in the various hoisting tests in HLSLFileCheck/hlsl/types/modifiers/const.

rkarrenberg commented on lib/HLSL/DxilValidation.cpp:
```diff
@@ -3308,6 +3312,10 @@ static void ValidateFunctionBody(Function *F, ValidationContext &ValCtx) {
         }
         if (IntegerType *IT = dyn_cast<IntegerType>(op->getType())) {
           if (IT->getBitWidth() == 8) {
+            // Allow i8* cast for llvm.lifetime.* intrinsics.
```
I agree, this doesn't make sense. Thanks!
Fixed.

rkarrenberg commented on lib/HLSL/DxilValidation.cpp:
```diff
@@ -3319,7 +3327,8 @@ static void ValidateFunctionBody(Function *F, ValidationContext &ValCtx) {
       if (IntegerType *IT = dyn_cast<IntegerType>(Ty)) {
-        if (IT->getBitWidth() == 8) {
+        // Allow i8* cast for llvm.lifetime.* intrinsics.
+        if (IT->getBitWidth() == 8 && !isa<BitCastInst>(I)) {
```
Fixed the first remark by using `onlyUsedByLifetimeMarkers`. Fixed second remark by querying shader model version at the top of ValidateFunctionBody and using it in two places.

rkarrenberg commented on lib/HLSL/DxilValidation.cpp:
```diff
@@ -3420,6 +3429,7 @@ static void ValidateFunctionBody(Function *F, ValidationContext &ValCtx) {
         BitCastInst *Cast = cast<BitCastInst>(&I);
         Type *FromTy = Cast->getOperand(0)->getType();
         Type *ToTy = Cast->getType();
+        if (ToTy == Type::getInt8PtrTy(ToTy->getContext())) continue; // Allow i8* cast for llvm.lifetime.* intrinsics.
```
Right, it's checking the same thing twice here. I still added the SM version check here, too, for clarity.

rkarrenberg commented on lib/Transforms/Scalar/HoistConstantArray.cpp:
```diff
@@ -250,15 +256,15 @@ bool CandidateArray::AllGEPUsersAreValid(GEPOperator *gep) {
 // Analyze all uses of the array to see if it qualifes as a constant array.
 // We check the following conditions:
-//  1. Make sure alloca is only used by GEP.
+//  1. Make sure alloca is only used by GEP and lifetime intrinsics.
```
I'm open to deleting the intrinsics, but I think they don't do any harm.
They only don't make sense if you assume that constant memory isn't allocated dynamically and can't be freed anyway. That's probably true for any back end, but you could also imagine another optimization phase undoing the hoisting again because it understands more context and can remove the memory alltogether. For instance, in one of the cases mentioned above, the global constant array actually ends up being 1.0f for every element - but further opt phases apparently fail to understand that, the entire loop with dynamic indexing could be removed using this information.

rkarrenberg commented on tools/clang/test/HLSLFileCheck/passes/hl/sroa_hlsl/memcpy_preuser.hlsl:
```diff
@@ -34,9 +36,9 @@ cbuffer cbuf : register(b1)
 float main(int doit : A) : SV_Target
 {
   float res = 0.0;
+  OuterStruct oStruct;
```
@pow2clk I'm slightly confused by the terminology "change" and "alteration" that you use, so apologies if I misunderstood you.
The original code could be replaced entirely by returning undef since the += always adds an uninitialized value. If lifetime intrinsics do not exist, the definition of oStruct is effectively hoisted outside the loop and suddenly the access is not 'undef' anymore but refers to the struct that's copied in the previous iteration of the loop. Of course that transformation is legal since undef can be replaced by an arbitrary value created from non-crashing, terminating code ;).

Now what's happening WITH lifetime intrinsics is not what I describe above: The opt pipeline actually fails to realize that there is an add with undef that it can use to further simplify the code. So a lot of stuff remains but the code looks different from before (and is likely to change if the opt pipeline becomes more aggressive with undef values). The resulting undef usage validation error stems from this, but I'd say that's more of a coincidence.

I believe that the test code should rather be modified to not exercise how much we can optimize undef values but instead do what was obviously originally intended with the i%2 == 0 conditional that refers to the previous iteration.

rkarrenberg replied:
For reference, here's the resulting DXIL of the modified test with lifetime intrinsics (can't attach a file :( ):
; ModuleID = 'D:\msft\DirectXShaderCompiler\tools\clang\test\HLSLFileCheck\passes\hl\sroa_hlsl\memcpy_preuser.hlsl'
target datalayout = "e-m:e-p:32:32-i1:32-i8:32-i16:32-i32:32-i64:64-f16:32-f32:32-f64:64-n8:16:32:64"
target triple = "dxil-ms-dx"

%struct.OuterStruct = type { float, float }
%cbuf = type { [1 x %struct.OuterStruct] }

define void @main() {
entry:
  %0 = call i32 @dx.op.loadInput.i32(i32 4, i32 0, i32 0, i8 0, i32 undef)
  %cmp.1 = icmp sgt i32 %0, 0
  br i1 %cmp.1, label %for.body.preheader, label %for.cond.cleanup

for.body.preheader:                               ; preds = %entry
  br label %for.body

for.cond.cleanup.loopexit:                        ; preds = %for.body
  br label %for.cond.cleanup

for.cond.cleanup:                                 ; preds = %for.cond.cleanup.loopexit, %entry
  %res.0 = phi float [ 0.000000e+00, %entry ], [ %add.res.1, %for.cond.cleanup.loopexit ]
  call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 0, float %res.0)
  ret void

for.body:                                         ; preds = %for.body.preheader, %for.body
  %res.1 = phi float [ %add.res.1, %for.body ], [ 0.000000e+00, %for.body.preheader ]
  %i.0 = phi i32 [ %inc, %for.body ], [ 0, %for.body.preheader ]
  %rem2 = and i32 %i.0, 1
  %cmp2 = icmp eq i32 %rem2, 0
  %add = fadd fast float %res.1, undef
  %add.res.1 = select i1 %cmp2, float %add, float %res.1
  %inc = add nuw nsw i32 %i.0, 1
  %exitcond = icmp eq i32 %inc, %0
  br i1 %exitcond, label %for.cond.cleanup.loopexit, label %for.body
}

; Function Attrs: nounwind readnone
declare i32 @dx.op.loadInput.i32(i32, i32, i32, i8, i32) #0

; Function Attrs: nounwind
declare void @dx.op.storeOutput.f32(i32, i32, i32, i8, float) #1

attributes #0 = { nounwind readnone }
attributes #1 = { nounwind }

!llvm.ident = !{!0}
!dx.version = !{!1}
!dx.valver = !{!2}
!dx.shaderModel = !{!3}
!dx.typeAnnotations = !{!4, !10}
!dx.viewIdState = !{!14}
!dx.entryPoints = !{!15}

!0 = !{!"clang version 3.7 (tags/RELEASE_370/final)"}
!1 = !{i32 1, i32 0}
!2 = !{i32 1, i32 6}
!3 = !{!"ps", i32 6, i32 0}
!4 = !{i32 0, %struct.OuterStruct undef, !5, %cbuf undef, !8}
!5 = !{i32 8, !6, !7}
!6 = !{i32 6, !"fval", i32 3, i32 0, i32 7, i32 9}
!7 = !{i32 6, !"fval2", i32 3, i32 4, i32 7, i32 9}
!8 = !{i32 8, !9}
!9 = !{i32 6, !"g_oStruct", i32 3, i32 0}
!10 = !{i32 1, void ()* @main, !11}
!11 = !{!12}
!12 = !{i32 0, !13, !13}
!13 = !{}
!14 = !{[3 x i32] [i32 1, i32 1, i32 1]}
!15 = !{void ()* @main, !"main", !16, null, null}
!16 = !{!17, !21, null}
!17 = !{!18}
!18 = !{i32 0, !"A", i8 4, i8 0, !19, i8 1, i32 1, i8 1, i32 0, i8 0, !20}
!19 = !{i32 0}
!20 = !{i32 3, i32 1}
!21 = !{!22}
!22 = !{i32 0, !"SV_Target", i8 9, i8 16, !19, i8 0, i32 1, i8 1, i32 0, i8 0, !20}

rkarrenberg replied:
And here's the DXIL of the modified test with lifetime intrinsics, which is basically identical to the original test without lifetimes:

%cbuf = type { [1 x %struct.OuterStruct] }
%struct.OuterStruct = type { float, float }
%dx.types.Handle = type { i8* }
%dx.types.CBufRet.f32 = type { float, float, float, float }

@cbuf = external constant %cbuf

define void @main() {
entry:
  %cbuf_cbuffer = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 2, i32 0, i32 1, i1 false)
  %0 = call i32 @dx.op.loadInput.i32(i32 4, i32 0, i32 0, i8 0, i32 undef)
  %cmp.1 = icmp sgt i32 %0, 0
  br i1 %cmp.1, label %for.body.preheader, label %for.cond.cleanup

for.body.preheader:                               ; preds = %entry
  br label %for.body

for.cond.cleanup.loopexit:                        ; preds = %for.body
  br label %for.cond.cleanup

for.cond.cleanup:                                 ; preds = %for.cond.cleanup.loopexit, %entry
  %res.0 = phi float [ 0.000000e+00, %entry ], [ %res.2, %for.cond.cleanup.loopexit ]
  call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 0, float %res.0)
  ret void

for.body:                                         ; preds = %for.body.preheader, %for.body
  %res.1 = phi float [ %res.2, %for.body ], [ 0.000000e+00, %for.body.preheader ]
  %oStruct.1.0 = phi float [ %2, %for.body ], [ undef, %for.body.preheader ]
  %i.0 = phi i32 [ %inc, %for.body ], [ 0, %for.body.preheader ]
  %rem2 = and i32 %i.0, 1
  %cmp2 = icmp eq i32 %rem2, 0
  %add = fadd fast float %oStruct.1.0, %res.1
  %res.2 = select i1 %cmp2, float %add, float %res.1
  %1 = call %dx.types.CBufRet.f32 @dx.op.cbufferLoadLegacy.f32(i32 59, %dx.types.Handle %cbuf_cbuffer, i32 %0)
  %2 = extractvalue %dx.types.CBufRet.f32 %1, 1
  %inc = add nuw nsw i32 %i.0, 1
  %exitcond = icmp eq i32 %inc, %0
  br i1 %exitcond, label %for.cond.cleanup.loopexit, label %for.body
}

pow2clk replied:
I recommend you use the three backticks "```" to quote code as it will annotate it and prevent accidentally mentioning github users as has happened here. I don't know why you were unable to attach files. That's a feature github should support.

I edited my original message to try to clarify that I'd performed the tests that you copied here already and what the outcome was.

The reason for the PR this test was made to validate was not to "optimize undef values", but to 1) allow memcpy replacements in circumstances other than this which would result in better optimized code and 2) avoid erroneous behavior in this case and possibly others similar to it we failed to conceive of. This test is as much to ensure that the previous behavior not be mistakenly restored as anything else.

I would like to know two things:
1) Why this PR causes an undef error with the original form of this hlsl file that wasn't produced before.
2) Why the additional CHECK lines were added as they are not newly produced with this PR and have no bearing on what this test is meant to evaluate.

rkarrenberg replied:
Ugh, very good point. Sorry for mentioning other people :(.
github said "Something went terribly wrong" when I tried uploading files ;).

1. The reason for the undef error is what I tried to explain above but obviously failed - let me try again :). The original test code makes use of an uninitialized value. However, if you don't have lifetime intrinsics that limit the scope of the struct to the loop body, the uninitialized value only exists in a loop phi node as incoming value from the preheader since the compiler doesn't know anymore that the struct definition is local and can never span multiple iterations. This is why it didn't manifests as an add with undef as one operand previously but now, with correct lifetime intrinsics, it does. 
2. I added the additional checks to ensure that there is a loop in the code. I'll gladly remove this again if you don't want this to be tested, it's indeed unrelated to the lifetime intrinsics.

tex3d replied:
Right, so my concern wasn't with the undef part exactly, that part makes perfect sense.  I was just concerned whether this test case would still repro the original issue that the test was meant to repro, without the memcpy change that went along with it.
However, It sounds like it may no longer be a good test case for this memcpy change anymore due to other recent changes made to SROA, if I'm interpreting this right.  So it may be testing some special case, or maybe just a more common one now.  I can let it go if Greg can :).

pow2clk replied:
Just to clarify, I was purely speculating that the recent SROA changes might have rendered this test moot. I'm about 10% sure that's the case.

More relevant, I'm about 90% (independent of the previous 10%) sure that moving this definition doesn't negatively impact the intent of this test.

rkarrenberg commented on lib/Transforms/Utils/PromoteMemoryToRegister.cpp:
```diff
@@ -562,6 +574,26 @@ void PromoteMem2Reg::run() {
+      if (!AnyStoreInStartBB) {
+        // No store in start BB. Add the lifetime marker as defining block to
+        // avoid undef.
+        Info.DefiningBlocks.insert(Info.DefiningBlocks.begin(), LifetimeStartBB);
```
1. This was not intentional, good catch! Fixed by including a "single store" condition in the guarding if statement.
2. I'll take a closer look at this tomorrow, and will also try to come up with a dedicated test.

rkarrenberg replied:
Alright, I refreshed my memory and did a bit of stepping through use cases of this.
I updated the code with a larger comment that explains what's going on in detail. Let me just quote the relevant code:
```
    // Calculate the set of read and write-locations for each alloca.  This is
    // analogous to finding the 'uses' and 'definitions' of each variable.
    Info.AnalyzeAlloca(AI);

    // Determine whether this alloca only has one store *before* potentially
    // adding a lifetime.start block as another definition. This allows to
    // rewrite such an alloca efficiently if possible but still benefit from
    // correct lifetime information should the single store not dominate all
    // loads.
    const bool IsSingleStoreAlloca = Info.DefiningBlocks.size() == 1;

    // For allocas that have at least one store, make sure the block that holds
    // the lifetime.start call is treated as a "definition". This is important
    // in cases where lifetime is restricted to a loop and not all loads are
    // dominated by stores. When walking up the CFG during ComputeLiveInBlocks,
    // this new "definition" prevents the value from being considered live in
    // in the loop header, which in turn prevents the value to be live across
    // the entire loop through a phi with undef as input from the preheader.
    // Ignore allocas that are only used in one block, even if the lifetime
    // start marker is in a different one, since those cases are trivial.
    if (LifetimeStartBB && !Info.DefiningBlocks.empty() && !Info.OnlyUsedInOneBlock) {
      bool AnyStoreInStartBB = false;
      for (BasicBlock *BB : Info.DefiningBlocks) {
        if (BB == LifetimeStartBB) {
          AnyStoreInStartBB = true;
          break;
        }
      }

      if (!AnyStoreInStartBB) {
        // No store in start BB. Add the lifetime marker as defining block.
        Info.DefiningBlocks.insert(Info.DefiningBlocks.begin(), LifetimeStartBB);
      }
    }
```

To answer your second point, the lifetime start block itself is added to `LiveInBlocks`, but its predecessors are not. This prevents ComputeLiveInBlocks from visiting blocks in which the lifetime of the object hasn't started. No load is required in any of the visited blocks for the algorithm to create phi nodes. And it's not checked again whether a defining block actually contains a store.
I think this is only relevant inside loops where previously the loop preheader would be visited even if the lifetime start block was inside the loop, causing a phi node in the header with undef coming from the preheader. This can later not be undone since the information where lifetime started has effectively been altered to a location outside the loop.

I added the following unit test which exercises exactly the two situations you mention. s.x is only stored in the conditional and loaded in the subsequent block, which isn't dominated by the store (thus `rewriteSingleStoreAlloca` fails, so we still need the LifetimeStartBB even if we hit that path - I've now allowed both things to coexist by checking `Info.DefiningBlocks.size() == 1` before adding the new def block). Without the fix, ComputeLiveInBlocks will add all blocks up to the entry to `LiveInBlocks`, and subsequently the IDF identifies the loop header as a location where `QueuePhiNode` has to insert a phi.
```
struct MyStruct {
  int x;
};

RWStructuredBuffer<MyStruct> g_rwbuf : register(u0);

void expensiveComputation(int i, int x)
{
  // We'd prefer this to not be inlined to better illustrate things but that's
  // not possible without more effort.
  g_rwbuf[i].x = x;
}

export
int conditional_struct_init_in_loop(int n, int c1)
{
  int res = n;

  for(int i=0; i<n; ++i) {
    expensiveComputation(i, i*c1); // s mut not be considered live here.

    MyStruct s;

    // Initialize struct conditionally.
    // NOTE: If some optimization decides to flatten the if statement or if the
    //       computation could be hoisted out of the loop, the phi with undef
    //       below will be replaced by the non-undef value (which is a valid
    //       "specialization" of undef).
    if (c1 < 0)
      s.x = g_rwbuf[i - c1].x;

    res = s.x; // i or undef.
  }

  return res; // Result is n if loop wasn't executed, n-1 if it was.
}
```

FWIW, I think it's really interesting that this is missing from LLVM even at top-of-tree, but maybe the issue doesn't appear anywhere as pronounced as in the DXR use case.

Sorry for the wall-of-text :).

rkarrenberg commented:
I'm working on file-check tests in parallel but haven't finished them. When I say "fixed" I mean locally, haven't pushed that yet. Would you prefer many small incremental commits for these fixes or that I collect them into a single update?

pow2clk commented:
I understood that is what "fixed" meant in this context. I prefer large commits.

Incidentally, I prefer bundled responses to comments which can be done by starting a review and responding to them as part of that "review" as well. It results in fewer email pings.

rkarrenberg commented on tools/clang/test/HLSLFileCheck/hlsl/control_flow/return/whole_scope_returned_loop.hlsl:
```diff
@@ -6,8 +6,6 @@ float main(float4 a:A) : SV_Target {
 // CHECK:%[[bReturned:.*]] = alloca i1
 // CHECK-NEXT:store i1 false, i1* %[[bReturned]]
-// Init retVal to 0.
-// CHECK:store float undef
```
"Init retVal to 0" looks like a stale comment to me, so I removed it: There's no retVal and no initialization to zero anywhere in that function.
There (strictly redundant) undef store is not generated anymore with the lifetime change.
However, this test does exhibit some more significant change in behavior that's related to lifetimes requiring a cleanup: The code ends up using an alloca i32 called "cleanup.dest.slot" that it stores to and uses in a switch at the end of the loop. This did not exist previously, but the test also doesn't check for/against it, so I'm not sure what to make of it.

rkarrenberg commented on tools/clang/test/HLSLFileCheck/samples/d3d11/ComputeShaderSort11.hlsl:
```diff
@@ -6,10 +6,10 @@
 // CHECK: addrspace(3)
 // CHECK: barrier
 // CHECK: addrspace(3)
+// CHECK: barrier
```
Makes sense. In this case it's just different code layout, two blocks have moved higher in the block list, nothing else. This does change the order of addrspace/barrier occurrences in the code, but then again there's many more occurrences of addrspace(3) than what's being checked here, so not sure how useful these checks are to begin with. There is a third barrier in the code even before my change, so maybe something that has changed since introduction of the test? Possibly worth doing this instead to prevent future regressions:
```
// CHECK: barrier
// CHECK: barrier
// CHECK: barrier
// CHECK-NOT: barrier
```

Here's the before/after IR:
```
define void @main() {
entry:
  %Data_UAV_structbuf = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 1, i32 0, i32 0, i1 false)  ; CreateHandle(resourceClass,rangeId,index,nonUniformIndex)
  %CB_cbuffer = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 2, i32 0, i32 0, i1 false)  ; CreateHandle(resourceClass,rangeId,index,nonUniformIndex)
  %0 = call i32 @dx.op.threadId.i32(i32 93, i32 0)  ; ThreadId(component)
  %1 = call i32 @dx.op.flattenedThreadIdInGroup.i32(i32 96)  ; FlattenedThreadIdInGroup()
  %2 = call %dx.types.ResRet.i32 @dx.op.bufferLoad.i32(i32 68, %dx.types.Handle %Data_UAV_structbuf, i32 %0, i32 0)  ; BufferLoad(srv,index,wot)
  %3 = extractvalue %dx.types.ResRet.i32 %2, 0
  %arrayidx = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %1
  store i32 %3, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  %4 = call %dx.types.CBufRet.i32 @dx.op.cbufferLoadLegacy.i32(i32 59, %dx.types.Handle %CB_cbuffer, i32 0)  ; CBufferLoadLegacy(handle,regIndex)
  %5 = extractvalue %dx.types.CBufRet.i32 %4, 0
  %shr = lshr i32 %5, 1
  %cmp.16 = icmp eq i32 %shr, 0
  br i1 %cmp.16, label %for.end, label %for.body.preheader

for.body.preheader:                               ; preds = %entry
  br label %for.body

for.body:                                         ; preds = %for.body, %for.body.preheader
  %j.0 = phi i32 [ %shr13, %for.body ], [ %shr, %for.body.preheader ]
  %neg = xor i32 %j.0, -1
  %and = and i32 %1, %neg
  %arrayidx2 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %and
  %6 = load i32, i32 addrspace(3)* %arrayidx2, align 4, !tbaa !25
  %or = or i32 %j.0, %1
  %arrayidx3 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %or
  %7 = load i32, i32 addrspace(3)* %arrayidx3, align 4, !tbaa !25
  %cmp4 = icmp ule i32 %6, %7
  %8 = call %dx.types.CBufRet.i32 @dx.op.cbufferLoadLegacy.i32(i32 59, %dx.types.Handle %CB_cbuffer, i32 0)  ; CBufferLoadLegacy(handle,regIndex)
  %9 = extractvalue %dx.types.CBufRet.i32 %8, 1
  %and6 = and i32 %9, %0
  %tobool7 = icmp ne i32 %and6, 0
  %cmp8tmp = xor i1 %cmp4, %tobool7
  %xor = xor i32 %j.0, %1
  %arrayidx10 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %xor
  %10 = load i32, i32 addrspace(3)* %arrayidx10, align 4, !tbaa !25
  %11 = load i32, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  %12 = select i1 %cmp8tmp, i32 %11, i32 %10
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  store i32 %12, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  %shr13 = lshr i32 %j.0, 1
  %cmp = icmp eq i32 %shr13, 0
  br i1 %cmp, label %for.end.loopexit, label %for.body

for.end.loopexit:                                 ; preds = %for.body
  br label %for.end

for.end:                                          ; preds = %for.end.loopexit, %entry
  %13 = load i32, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.bufferStore.i32(i32 69, %dx.types.Handle %Data_UAV_structbuf, i32 %0, i32 0, i32 %13, i32 undef, i32 undef, i32 undef, i8 1)  ; BufferStore(uav,coord0,coord1,value0,value1,value2,value3,mask)
  ret void
}
```

```
define void @main() {
entry:
  %Data_UAV_structbuf = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 1, i32 0, i32 0, i1 false)  ; CreateHandle(resourceClass,rangeId,index,nonUniformIndex)
  %CB_cbuffer = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 2, i32 0, i32 0, i1 false)  ; CreateHandle(resourceClass,rangeId,index,nonUniformIndex)
  %0 = call i32 @dx.op.threadId.i32(i32 93, i32 0)  ; ThreadId(component)
  %1 = call i32 @dx.op.flattenedThreadIdInGroup.i32(i32 96)  ; FlattenedThreadIdInGroup()
  %2 = call %dx.types.ResRet.i32 @dx.op.bufferLoad.i32(i32 68, %dx.types.Handle %Data_UAV_structbuf, i32 %0, i32 0)  ; BufferLoad(srv,index,wot)
  %3 = extractvalue %dx.types.ResRet.i32 %2, 0
  %arrayidx = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %1
  store i32 %3, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  %4 = call %dx.types.CBufRet.i32 @dx.op.cbufferLoadLegacy.i32(i32 59, %dx.types.Handle %CB_cbuffer, i32 0)  ; CBufferLoadLegacy(handle,regIndex)
  %5 = extractvalue %dx.types.CBufRet.i32 %4, 0
  %shr = lshr i32 %5, 1
  %cmp.16 = icmp eq i32 %shr, 0
  br i1 %cmp.16, label %for.cond.cleanup, label %for.body.preheader

for.body.preheader:                               ; preds = %entry
  br label %for.body

for.cond.cleanup.loopexit:                        ; preds = %for.body
  br label %for.cond.cleanup

for.cond.cleanup:                                 ; preds = %for.cond.cleanup.loopexit, %entry
  %6 = load i32, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.bufferStore.i32(i32 69, %dx.types.Handle %Data_UAV_structbuf, i32 %0, i32 0, i32 %6, i32 undef, i32 undef, i32 undef, i8 1)  ; BufferStore(uav,coord0,coord1,value0,value1,value2,value3,mask)
  ret void

for.body:                                         ; preds = %for.body, %for.body.preheader
  %j.0 = phi i32 [ %shr13, %for.body ], [ %shr, %for.body.preheader ]
  %neg = xor i32 %j.0, -1
  %and = and i32 %1, %neg
  %arrayidx2 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %and
  %7 = load i32, i32 addrspace(3)* %arrayidx2, align 4, !tbaa !25
  %or = or i32 %j.0, %1
  %arrayidx3 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %or
  %8 = load i32, i32 addrspace(3)* %arrayidx3, align 4, !tbaa !25
  %cmp4 = icmp ule i32 %7, %8
  %9 = call %dx.types.CBufRet.i32 @dx.op.cbufferLoadLegacy.i32(i32 59, %dx.types.Handle %CB_cbuffer, i32 0)  ; CBufferLoadLegacy(handle,regIndex)
  %10 = extractvalue %dx.types.CBufRet.i32 %9, 1
  %and6 = and i32 %10, %0
  %tobool7 = icmp ne i32 %and6, 0
  %cmp8tmp = xor i1 %cmp4, %tobool7
  %xor = xor i32 %j.0, %1
  %arrayidx10 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %xor
  %11 = load i32, i32 addrspace(3)* %arrayidx10, align 4, !tbaa !25
  %12 = load i32, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  %13 = select i1 %cmp8tmp, i32 %12, i32 %11
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  store i32 %13, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  %shr13 = lshr i32 %j.0, 1
  %cmp = icmp eq i32 %shr13, 0
  br i1 %cmp, label %for.cond.cleanup.loopexit, label %for.body
}
```

tex3d replied:
Is there anything that can be done to fix the ordering of loopexit and cleanup before the loop body with this change?

tex3d replied:
Ok, I think we can live with this for now then.

rkarrenberg commented:
I somehow don't have the option to create a review when I'm commenting on existing comments. Unless there's some other option to batch comments that I'm missing I have to stick to sending one mail per comment. Sorry for that :(.

rkarrenberg commented:
> 
> ❌ [Build DirectXShaderCompiler 1.0.3432 failed](https://ci.appveyor.com/project/antiagainst/directxshadercompiler/builds/34086474) (commit [a86abe5](https://github.com/microsoft/DirectXShaderCompiler/commit/a86abe58b0) by @rkarrenberg)

Not sure what went wrong here, the new unit test executes fine locally :(.

pow2clk commented on tools/clang/lib/CodeGen/CGCleanup.cpp:
```diff
@@ -942,15 +942,18 @@ bool CodeGenFunction::isObviouslyBranchWithoutCleanups(JumpDest Dest) const {
   // Create the branch.
-  llvm::BranchInst *BI = Builder.CreateBr(Dest.getBlock());
+  // HLSL Change Begin - use pre-generated branch if exists
+  // If not a wave-enabled stage, we can keep everything unconditional as before
```
I'm not sure the last line of this comment makes sense here. Even in a non-wave stage, WaveBr should be non-null.

pow2clk commented on tools/clang/lib/CodeGen/CGStmt.cpp:
```diff
@@ -1202,11 +1202,19 @@ void CodeGenFunction::EmitBreakStmt(const BreakStmt &S) {
+
+    // Insertion of lifetime.start/end intrinsics may require a cleanup, so we
+    // pass the branch that we already generated into the handler.
+    EmitBranchThroughCleanup(BreakContinueStack.back().BreakBlock, waveBr);
```
As this is probably the only part of this change where I may have particular expertise, I feel compelled to affirmatively say that this looks good to me 😄

pow2clk reviewed (commented):
I skipped the tests this time.  I'm still puzzled by why some of the output is moved around.

Mostly nitpicky stuff. I feel pretty good about it.

tex3d commented on lib/HLSL/DxilPreparePasses.cpp:
```diff
@@ -376,6 +376,96 @@ class DxilFinalizeModule : public ModulePass {
+        if (ValMajor < 1 || (ValMajor == 1 && ValMinor < 6))
+          IRBuilder<>(CI).CreateStore(Constant::getNullValue(T), ptr);
+        else
+          IRBuilder<>(CI).CreateStore(UndefValue::get(T), ptr);
```
Note: Still haven't confirmed whether undef store is well tolerated by existing drivers.

tex3d commented on lib/Transforms/Scalar/DxilLoopUnroll.cpp:
```diff
@@ -522,8 +523,16 @@ static bool BreakUpArrayAllocas(bool AllowOOBIndex, IteratorT ItBegin, IteratorT
+      // Ignore lifetime intrinsics
+      if (BitCastInst *BCI = dyn_cast<BitCastInst>(U)) {
+        DXASSERT(onlyUsedByLifetimeMarkers(BCI),
+                 "expected bitcast to only be used by lifetime intrinsics");
```
Is there no possibility that bitcast for memcpy could be present at this point, or something else?

tex3d reviewed (changes requested):
See some prior comments.

Also, I thought we were going to have an execution test added to ExecutionTest.cpp for when lifetime markers aren't removed by mem2reg.  It would need to test 6 driver scenarios:
1. regular shader target (like compute), SM 6.0 and -validator-version 1.5 (or below), to test undef store case
2. library target, lib_6_3 and -validator-version 1.5 (or below), to test undef store case
3. regular shader target, SM 6.0 and -validator-version 1.6, to test zeroinitializer store case
4. library target, lib_6_3 and -validator-version 1.6, to test zeroinitializer store case
5. regular shader target, SM 6.6, to test lifetime markers
6. library target, lib_6_6, to test lifetime markers

rkarrenberg reviewed (commented):
Thanks for the comments Tex and Greg! They all made sense, and all locations should be fixed in the next commit.

Next, I'll be working on execution tests as requested by Tex.

Unfortunately, merging the latest master into the lifetimes branch has uncovered an issue with commit 5d6674ad0a19a3ad7c3ea470197f3a73443da979 ("Hoist exits out of loops when unrolling to make code structured" by @adam-yang) and the tests it added. I've spent the best of two days to ensure that the lifetimes did not actually break anything. The conclusion unfortunately is that the structurization pass has some kind of a phase ordering problem, i.e., it depends on a certain "normalization" of the code (I think mostly because of a current restriction that all values used in an exit block must dominate the loop latch).
The cleanup code blocks that are introduced for lifetime intrinsics have an impact on this because DxilLoopUnroll runs relatively early in the pipeline. I've experimented with adding some passes right in front (InstCombine, SimplifyCFG, JumpThreading), which fixes the situation and allows the structurization to kick in in more cases, but while that fixes the dedicated tests, others start breaking, and I'm not sure I should go down that rabbit hole.
I verified that stock LLVM is generating exactly the same code as dxc with lifetimes enabled. In basically all cases I tried (starting from the structurize tests) the code was eventually cleaned up so that structurize would be able to kick in if it was run later. I guess that's not desired, though?

Side note: When investigating the structurize problem I stumbled over a bug in `CodeMetrics::analyzeBasicBlock` (called by `ApproximateLoopSize`) which treats calls to things such as dx.op.annotateHandle as actual calls, which messes with the heuristics.

rkarrenberg commented:
Oh, that's unfortunate... I explicitly wanted to *not* merge before the fixes because of the issue with DxilLoopUnroll mentioned above.

pow2clk commented on tools/clang/unittests/HLSL/ExecutionTest.cpp:
```diff
@@ -1272,6 +1283,263 @@ void ExecutionTest::RunRWByteBufferComputeTest(ID3D12Device *pDevice, LPCSTR pSh
+
+  // Test library with lifetime intrinsics.
+  RunLifetimeIntrinsicTest(pDevice, pShader, D3D_SHADER_MODEL_6_5, true, pOptions16, _countof(pOptions16), values); // TODO: Test 6.6 here!
+  VERIFY_ARE_EQUAL(values[0], (uint32_t)0);
```
This ExecutionTest looks really good and I appreciate how easy it is to read.

The only suggestion I'd make is to not rely on zero as an indicator of success. So many failures can result in memory being zero which might lead to a false positive, I would recommend avoiding it. 

Even something only slightly more complex like comparing `values[1]` to `1` would make me feel better.

rkarrenberg replied:
Thanks Greg! I've updated the test with your suggestion.
In general, the test result doesn't really matter. The test is only about the runtime/driver accepting the generated code (store zeroinitializer, store undef, or lifetime intrinsics depending on the SM/validator version). I saw the pattern of testing against value 0 in the test from which I copied, so just stuck with that :).

adam-yang commented on lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp:
```diff
@@ -3206,6 +3256,12 @@ static bool ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC,
+  // Unless we encounter an issue, we always replace V by Src below.
+  // Conservatively remove all lifetime users of V. Otherwise, Src would inherit
+  // them from V and suddenly have a different lifetime than it had before.
+  removeLifetimeUsers(V);
```
If V is read outside of Src's lifetime (V's lifetime extends beyond Src's lifetime), the lifetime of Src needs to be expanded to include V.

rkarrenberg replied:
Hmm, the intention of this was to be conservative. I think I experimented with expanding lifetimes but hit some blockers.

If the code below actually performs the replacement of Src by V then removing lifetime intrinsics of V should be fine. The intrinsics of Src are dead code afterwards (they still exist, but there are no uses of the memory that they refer to anymore). So the result is that neither memory described by Src nor V has lifetime intrinsics attached afterwards, which is conservative.

Now that leaves the case where the code below ends up NOT doing the replacement. IIUC there's only two cases where this happens:
1. if V is not a constant but V == Src
2. if V is not a constant and the memcpy is not an unbounded array memcpy and Src and V are pointers to structs of identical layouts and either Src or V has more than one use.

Again, removing lifetime intrinsics of V should be conservative here. Since both V and Src are still separate this should not have any effect on Src. If lifetimes of Src are somehow wrong then the problem was there before already.

I must be missing something...

adam-yang replied:
I see what the problem is.

V is actually the one being replaced by Src, not the other way around. `removeLifetimeUsers` removed all of V's lifetime markers, but when V is removed and all its usage is replaced by Src. Now Src's lifetime markers stick around and are wrong.

But even if we do `removeLifetimeUsers` on Src instead, by doing `replaceAllUsesWith`, V's lifetime markers are now pointing to Src, which is still wrong. To achieve the intended result, we must remove the lifetime markers for both pointers.

But removing it here is maybe overly conservative, not all the cases below replace the pointers. Here's a patch:
```
diff --git a/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp b/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp
index 33ca05ee3..70a081ad6 100644
--- a/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp
+++ b/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp
@@ -3237,6 +3237,13 @@ static void removeLifetimeUsers(Value *V) {
   }
 }
 
+// Should be merging the lifetime, but for convenience, just
+// remove them now to be safe.
+static void updateLifetimeForReplacement(Value *From, Value *To) {
+  removeLifetimeUsers(From);
+  removeLifetimeUsers(To);
+}
+
 static bool DominateAllUsers(Instruction *I, Value *V, DominatorTree *DT);
 
 
@@ -3257,14 +3264,10 @@ static bool ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC,
     if (!DominateAllUsers(SrcI, V, DT))
       return false;
 
-  // Unless we encounter an issue, we always replace V by Src below.
-  // Conservatively remove all lifetime users of V. Otherwise, Src would inherit
-  // them from V and suddenly have a different lifetime than it had before.
-  removeLifetimeUsers(V);
-
   Type *TyV = V->getType()->getPointerElementType();
   Type *TySrc = Src->getType()->getPointerElementType();
   if (Constant *C = dyn_cast<Constant>(V)) {
+    updateLifetimeForReplacement(V, Src);
     if (TyV == TySrc) {
       if (isa<Constant>(Src)) {
         V->replaceAllUsesWith(Src);
@@ -3280,8 +3283,10 @@ static bool ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC,
     }
   } else {
     if (TyV == TySrc) {
-      if (V != Src)
+      if (V != Src) {
+        updateLifetimeForReplacement(V, Src);
         V->replaceAllUsesWith(Src);
+      }
     } else if (!IsUnboundedArrayMemcpy(TyV, TySrc)) {
       Value* DestVal = MC->getRawDest();
       Value* SrcVal = MC->getRawSource();
@@ -3322,6 +3327,7 @@ static bool ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC,
             MemcpySplitter::SplitMemCpy(MC, DL, annotation, typeSys);
             return true;
           } else {
+            updateLifetimeForReplacement(V, Src);
             DstPtr->replaceAllUsesWith(SrcPtr);
           }
         } else {
@@ -3330,6 +3336,7 @@ static bool ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC,
         }
       }
     } else {
+      updateLifetimeForReplacement(V, Src);
       DXASSERT(IsUnboundedArrayMemcpy(TyV, TySrc), "otherwise mismatched types in memcpy are not unbounded array");
       ReplaceUnboundedArrayUses(V, Src);
     }
```

rkarrenberg replied:
Oh, I was getting confused by what is being replaced by what, sorry for that.
Yes, you are right, contrary to my comment in the code, removing only the lifetimes of V is not conservative. Replacing the memcpy effectively merges memory accesses of two previously disjoint allocas into a single alloca. If the allocas previously had non-equal lifetimes, leaving the lifetimes of Src intact can be wrong just the same way as leaving those of V intact (which apparently happens more frequently since I did encounter this case).
I also agree with your patch. It's a bit ugly to have multiple places where we remove lifetimes, but it's clearly better than unconditionally removing them, and I prefer this over restructuring the rest of that code to pull the cases that don't actually replace anything out.
I'll apply the patch and will also add a test for this.

tex3d commented:
I have made a minimal example showing how using lifetime markers in intermediate code in the compiler makes it produce extra live values in loops, resulting in performance regressions in real shaders, even when lifetime markers are eliminated at the first mem2reg.

```
// RUN: %dxc -T lib_6_6 %s | FileCheck %s

// CHECK-NOT: undef

int loopCount;

bool done(int);
float loop_code();

void fn(out float res) {
  for (int i = 0; i < loopCount; i++) {
    float f = loop_code();
    if (done(i)) {
      res = f;
      return;
    }
  }
  res = 1;
}

export
void main(out float res : OUT) {
  res = 0;
  float f;
  fn(f);
  if (f > 0)
    res = f;
}
```

Block without lifetimes markers generated during CodeGen (with and without -Zi for convenience):
```
"\01?fn@@YAXAIAM@Z.exit":                         ; preds = %"\01?fn@@YAXAIAM@Z.exit.loopexit", %entry
  %f.0 = phi float [ 1.000000e+00, %entry ], [ %f.0.ph, %"\01?fn@@YAXAIAM@Z.exit.loopexit" ]
  call void @llvm.dbg.value(metadata float %f.0, i64 0, metadata !71, metadata !46), !dbg !70 ; var:"f" !DIExpression()
  call void @llvm.dbg.value(metadata float %f.0, i64 0, metadata !72, metadata !46), !dbg !73 ; var:"res" !DIExpression()
  %cmp = fcmp fast ogt float %f.0, 0.000000e+00, !dbg !74 ; line:24 col:9
  br i1 %cmp, label %if.then, label %if.end, !dbg !76 ; line:24 col:7

; No -Zi:
; <label>:18                                      ; preds = %16, %0
  %19 = phi float [ 1.000000e+00, %0 ], [ %17, %16 ]
  %20 = fcmp fast ogt float %19, 0.000000e+00
  br i1 %20, label %21, label %22
```

Same block with lifetime markers generated during CodeGen:
```
cleanup.3.i:                                      ; preds = %cleanup.3.i.loopexit, %entry
  %f.3 = phi float [ undef, %entry ], [ %call.i, %cleanup.3.i.loopexit ]
  %cleanup.dest.slot.i.1 = phi i32 [ 2, %entry ], [ %cleanup.dest.slot.i.1.ph, %cleanup.3.i.loopexit ]
  call void @llvm.dbg.value(metadata float %f.3, i64 0, metadata !59, metadata !45), !dbg !58 ; var:"f" !DIExpression()
  call void @llvm.dbg.value(metadata float %f.3, i64 0, metadata !60, metadata !45), !dbg !61 ; var:"res" !DIExpression()
  %switch.i = icmp eq i32 %cleanup.dest.slot.i.1, 2, !dbg !72 ; line:23 col:3
  call void @llvm.dbg.value(metadata float 1.000000e+00, i64 0, metadata !60, metadata !45), !dbg !61 ; var:"res" !DIExpression()
  call void @llvm.dbg.value(metadata float 1.000000e+00, i64 0, metadata !59, metadata !45), !dbg !58 ; var:"f" !DIExpression()
  %.f.3 = select i1 %switch.i, float 1.000000e+00, float %f.3, !dbg !72 ; line:23 col:3
  call void @llvm.dbg.value(metadata float %.f.3, i64 0, metadata !59, metadata !45), !dbg !58 ; var:"f" !DIExpression()
  call void @llvm.dbg.value(metadata float %.f.3, i64 0, metadata !60, metadata !45), !dbg !61 ; var:"res" !DIExpression()
  %cmp = fcmp fast ogt float %.f.3, 0.000000e+00, !dbg !74 ; line:24 col:9
  br i1 %cmp, label %if.then, label %if.end, !dbg !76 ; line:24 col:7

; No -Zi:
; <label>:19                                      ; preds = %17, %0
  %20 = phi float [ undef, %0 ], [ %9, %17 ]
  %21 = phi i32 [ 2, %0 ], [ %18, %17 ]
  %22 = icmp eq i32 %21, 2
  %23 = select i1 %22, float 1.000000e+00, float %20
  %24 = fcmp fast ogt float %23, 0.000000e+00
  br i1 %24, label %25, label %26
```

The problem is that the cleanup block that's introduced during codegen creates a convergence point in the CFG from which it may choose to continue looping, or end the loop.  All loop exits exit pass through the cleanup block, so the compiler can no longer keep values assigned before exiting the loop separate from values running around the loop.  There is also an extra phi to track the cleanup dest slot so it can select the right value at exit.  Of course the amount of extra data can be much larger than the minimal example illustrates.  This issue causes performance regressions in shaders on current shader models in our testing.

rkarrenberg commented:
> I have made a minimal example showing how using lifetime markers in intermediate code in the compiler makes it produce extra live values in loops, resulting in performance regressions in real shaders, even when lifetime markers are eliminated at the first mem2reg.

> The problem is that the cleanup block that's introduced during codegen creates a convergence point in the CFG from which it may choose to continue looping, or end the loop. All loop exits exit pass through the cleanup block, so the compiler can no longer keep values assigned before exiting the loop separate from values running around the loop. There is also an extra phi to track the cleanup dest slot so it can select the right value at exit. Of course the amount of extra data can be much larger than the minimal example illustrates. This issue causes performance regressions in shaders on current shader models in our testing.

Tex, apparently I can't reply to your comment directly, sorry for that.

Thanks for that example, that's super useful! I dug into it and realized this was basically the same thing that happened in combination with the loop structurization pass.
I modified the example a bit so I can also feed it as C++ to clang (no constant buffer accesses etc.):
```
// For the C++ variant, remove `in`, replace `out float` by `float*`, pass `&f` to fn2, etc.
void fn2(in int loopCount, out float res) {
  for (int i = 0; i < loopCount; i++) {
    float f = loop_code();
    if (done(i)) {
      res = f;
      return;
    }
    if (done(-i)) {
      res = 2;
      return;
    }
  }
  res = 1;
}

export
void test2(in int loopCount, out float res) {
  float f;
  fn2(loopCount, f);
  res = f;
}
```

I then checked with stock LLVM, which produces a clean result as expected (and has lifetimes enabled, of course).
The minimal set of optimizations required for optimal output was this:
```
- clang++ -fno-discard-value-names -emit-llvm -g0 -O0 -o x.ll
- remove optnone and noinline attributes from file, add alwaysinline to fn2
- opt -S -always-inline -sroa -simplifycfg -loop-rotate -jump-threading -simplifycfg -instcombine x.ll -o y.ll
https://godbolt.org/z/YoPcdT
```

Now based on that I ran opt on the output of dxc and this minimal set of optimizations cleaned up the code, too:
```
- dxc -T lib_6_6 z.hlsl > z.ll // code from above
- remove target triple from z.ll
- opt -jump-threading -simplifycfg -instcombine -S z.ll -o w.ll
```

So that tells us that the problem is most likely not with lifetime and cleanup code. Stock LLVM generates similar intermediate code the one produced by dxc with cleanup blocks and phi nodes with constants that note which exit was taken. It's just cleaned up better. Running a small set of optimizations on dxc's final output also cleans it up. The question is now: Is it possible to add such cleanup passes somewhere, or do we have to prevent this code from being generated in the first place. I'm not sure if the latter is feasible...

Now here's another problem. I added the following to BackendUtil.cpp at the very end of `EmitAssembly()` to mimic what I did with opt:
```
  legacy::PassManager PM;
  PM.add(createJumpThreadingPass());
  PM.add(createCFGSimplificationPass());
  PM.add(createInstructionCombiningPass());
  PM.run(*TheModule);
```
When I dump the code before running the pass manager, write it to a file, and run `opt -jump-threading -simplifycfg -instcombine`, the code is cleaned up. Running through the pass manager, however, there is basically no difference. Stock LLVM's CFG simplification and jump threading have a lot more logic than the versions in dxc, and apparently the ones in dxc fail to clean it up :disappointed:.

Next interesting experiment: I enabled lifetimes and disabled all the special sauce marked with "HLSL Change" in PassManagerBuilder.cpp. This obviously doesn't produce final code, but I can still dump at the end of EmitAssembly() - and indeed, the code is clean, there's no undef phis, and no cleanup blocks left. In fact, the code is even cleaner than what dxc TOT produces, thread jumping removed one more unnecessary branch.

This all leads me to believe that we're dealing with an inefficiency in the optimization phase ordering. Unfortunately, even running basically the entire pass pipeline again on the final code didn't clean it up, so it's a combination of passes not being powerful enough (vs stock LLVM) and some unfortunate ordering of transformations (since running just the non-HLSL stuff does manage the cleanup). Not sure where to continue... @tex3d @pow2clk @adam-yang what do you guys think?

rkarrenberg commented:
Next update:
I managed to clean up the code by adding SROA and jump threading passes at the start of `populateModulePassManager()`:
```
void PassManagerBuilder::populateModulePassManager(
    legacy::PassManagerBase &MPM) {

+  MPM.add(createSROAPass());
+  MPM.add(createJumpThreadingPass());

  // If all optimizations are disabled, just run the always-inline pass and,
  // if enabled, the function merging pass.
  if (OptLevel == 0) {
```

This produced clean code for all lifetime cases I had, but regressed `hcttest clang`:
```
Summary of Non-passing Tests (showing 20 of 43):
    PixTest::CompileDebugLines [Failed]
    DxilModuleTest::Precise4 [Failed]
    CompilerTest::BatchHLSL [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\bitpiece_full_var.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\dbg_value_phi_loop.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\locals\temporary_dbg_declare.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\mat3x2_dbg.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\hlsl\compile_options\Qstrip_reflect\Qstrip_reflect_struct_buf.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\mat_dbg.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\nested_struct.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\new_noops_call.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\new_noops_groupshare.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\new_noops_mandatory_immed.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\new_noops_mandatory_immed_load.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\new_noops_memcpy2.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\new_noops_no_fold.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\new_noops_no_fold_vec.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\new_noop_no_fold_double.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\new_noop_no_fold_int.hlsl [Failed]
        D:\msft\DirectXShaderCompiler\tools\clang\test\HLSL\..\HLSLFileCheck\dxil\debug\new_noop_resource_var.hlsl [Failed]
```
Tex, can you try that locally and see if it fixes the internal regression(s)? Maybe it makes sense to look at those 43 failures...

pow2clk commented:
I looked over the most recent changes. They look good. @tex3d and @adam-yang, can you look over your requested changes and verify this is ready to go?

tex3d commented on include/dxc/Support/HLSLOptions.td:
```diff
@@ -271,6 +271,10 @@ def validator_version : Separate<["-", "/"], "validator-version">, Group<hlslcom
   HelpText<"Override validator version for module.  Format: <major.minor> ; Default: DXIL.dll version or current internal version.">;
 def print_after_all : Flag<["-", "/"], "print-after-all">, Group<hlslcomp_Group>, Flags<[CoreOption, HelpHidden]>,
   HelpText<"Print LLVM IR after each pass.">;
+def force_zero_store_lifetimes : Flag<["-", "/"], "force-zero-store-lifetimes">, Group<hlslcomp_Group>, Flags<[CoreOption]>,
```
I think this option should be HelpHidden, but don't hold the PR for this.  I don't think we intend to keep this option long-term, it's a testing option while we decide whether undef stores are safe for downlevel.

tex3d commented on tools/clang/test/HLSLFileCheck/passes/llvm/simplifycfg/fold-cond-branch-on-phi.hlsl:
```diff
@@ -7,6 +7,7 @@
 // CHECK: %[[cond:.+]] = phi i1
 // CHECK-SAME: [ false
 // CHECK: br i1 %[[cond]]
+// CHECK: @main
```
I see one added check: for `@main` after the checks for things tat should be *in* main, which is strange.  This test uses /Zi, so it should include the special CHECK-LABEL after IR checks to make sure you're not matching strings from the included source file:
```
// Exclude quoted source file (see readme)
// CHECK-LABEL: {{!"[^"]*\\0A[^"]*"}}
```

pow2clk replied:
As it happens, the previous checks won't match the debug inclusions because they include regexp, which is why this shader didn't have it before. I don't know why ralf added the `@main` check, but it's often just a check that nothing failed and the debug info that would match `@main` won't be produced unless a successful compile has occurred. 

Adding the suggested CHECK-LABEL might be a good precaution regardless, but the readme it references has been removed so it's probably more confusing to reference it than not.

tex3d replied:
The readme that was removed used to be hard to find, as it was nested under some debug sub-folder, and there were other readme files that weren't the one you were looking for.  Since then, the text explaining this was moved into the top-level Readme.md under the HLSLFileCheck directory, so it should be much easier to find.

I too do not know why this `@main` check was added here in the first place.  My worry was that the @main check might only match because it was matching a sub-string in the included file string, since it should be located before the instructions in the function, right?

pow2clk replied:
Good to know! I thought it was just gone.

I share your worry and await consolation.

rkarrenberg replied:
Damn it... I overlooked this one, failed to properly revert it :(.
It was only added previously when I had the rest of the checks disabled because tests without any CHECK fail.

pow2clk replied:
No worries. If you want to make a quick fix change, I promise it won't take as long as this one did 😉

rkarrenberg replied:
No problem, but I need some advice how to submit the fix. A new PR? Or simply push a new commit to this one and you guys will merge the lifetimes branch again?

pow2clk replied:
A new PR would be simplest

rkarrenberg replied:
Here you go :).
https://github.com/microsoft/DirectXShaderCompiler/pull/3264
For future reference, could I have reused the same 'lifetimes' branch on my fork instead of creating a new branch? I was unsure, so went for the "safe" variant.

pow2clk replied:
Once this PR was completed, you could use the same branch to create another PR.

tex3d reviewed (approved):
Looks good.  Note a couple minor comments that should not block merging, and can be cleaned up later.

adam-yang reviewed (approved):
Approved. Thanks for working out the structurize loop exit and memcpy issue!

pow2clk commented:
🥳
